### PR TITLE
adding first & granted steps previousProtocol 4516

### PR DESCRIPTION
**Pull request by parvinba**

### CODE REVIEW

**Review by mohomeoffice — APPROVED** [review]
it seems okay to me

File: pages/project-version/middleware/index.js [hunk 1]
```diff
@@ -304,9 +304,13 @@ const getPreviousProtocols = (firstVersion, previousVersion, grantedVersion) =>
   const previous = get(previousVersion, 'data.protocols', []).filter(Boolean).filter(p => !p.deleted).map(p => p.id);
   const granted = get(grantedVersion, 'data.protocols', []).filter(Boolean).map(p => p.id);
   const showDeleted = uniq([ ...previous, ...granted ]);
-  const allSteps = get(previousVersion, 'data.protocols', []).filter(Boolean).filter(p => !p.deleted).map(p => p.steps);
+  let allSteps = get(previousVersion, 'data.protocols', []).filter(Boolean).filter(p => !p.deleted).map(p => p.steps);
```
Comment (jeff-horton-ho-sas) [inline pages/project-version/middleware/index.js:307]: Reusing allStep is confusing,and theres a lot of repeated logic here. I'd suggest renaming to previous/first steps, tbh I think it's clearer if these are extracted to a function:
```suggestion
  function extractActiveSteps(version) {
    return get(version, 'data.protocols', [])
      .filter(Boolean)
      .filter(p => !p.deleted)
      .map(p => p.steps)
      .map((element) => element && Array.isArray(element) ? element.filter(s => !s.deleted) : element);
  }
  
  const steps = extractActiveSteps(previousVersion);
  const firstSteps = extractActiveSteps(firstVersion);
  const grantedSteps = extractActiveSteps(grantedVersion);
```

**Review by jeff-horton-ho-sas — CHANGES_REQUESTED** [review]

File: pages/project-version/middleware/index.js [hunk 2]
```diff
@@ -304,9 +304,23 @@ const getPreviousProtocols = (firstVersion, previousVersion, grantedVersion) =>
...
+  
+  const steps = extractActiveSteps(previousVersion);
+  const firstSteps = extractActiveSteps(firstVersion);
+  const grantedSteps = extractActiveSteps(grantedVersion);
   const steps = allSteps.map((element) => element && Array.isArray(element) ? element.filter(s => !s.deleted) : element);
-  return { first, previous, granted, showDeleted, steps };
+  allSteps = get(firstVersion, 'data.protocols', []).filter(Boolean).filter(p => !p.deleted).map(p => p.steps);
+  const firstSteps = allSteps.map((element) => element && Array.isArray(element) ? element.filter(s => !s.deleted) : element);
+  allSteps = get(grantedVersion, 'data.protocols', []).filter(Boolean).filter(p => !p.deleted).map(p => p.steps);
+  const grantedSteps = allSteps.map((element) => element && Array.isArray(element) ? element.filter(s => !s.deleted) : element);
```
Comment (jeff-horton-ho-sas) [inline pages/project-version/middleware/index.js:322]: This is now redundant due to change above (sorry for the confusion - the GitHub UI doesn't let you ihave suggestions that span deleted lines)

```suggestion
  allSteps = get(firstVersion, 'data.protocols', []).filter(Boolean).filter(p => !p.deleted).map(p => p.steps);
  const firstSteps = allSteps.map((element) => element && Array.isArray(element) ? element.filter(s => !s.deleted) : element);
  allSteps = get(grantedVersion, 'data.protocols', []).filter(Boolean).filter(p => !p.deleted).map(p => p.steps);
  const grantedSteps = allSteps.map((element) => element && Array.isArray(element) ? element.filter(s => !s.deleted) : element);
```

**Review by jeff-horton-ho-sas — APPROVED** [review]

File: pages/project-version/middleware/index.js (hunk 2 above)
Comment (parvinba) [reply pages/project-version/middleware/index.js]: Thanks.. I already did that..in latest commit.. Thanks for all suggestions..

**Pull request merged**